runtime/commonMain/src/kotlinx/serialization/Tagged.kt

```diff
@@ -64,17 +64,27 @@ abstract class TaggedEncoder<Tag : Any?> : Encoder, CompositeEncoder {
 
     // ---- Implementation of low-level API ----
 
-    fun encodeElement(desc: SerialDescriptor, index: Int): Boolean {
+    fun <T: Any?> encodeElement(desc: SerialDescriptor, index: Int, value: T): Boolean {
         val tag = desc.getTag(index)
-        val shouldWriteElement = shouldWriteElement(desc, tag, index)
+        val shouldWriteElement = shouldWriteElement(desc, tag, index, value)
+        if (shouldWriteElement) {
+            pushTag(tag)
+        }
+        return shouldWriteElement
+    }
+
+    fun encodeNullElement(desc: SerialDescriptor, index: Int): Boolean {
+        val tag = desc.getTag(index)
+        val shouldWriteElement = shouldWriteNullElement(desc, tag, index)
         if (shouldWriteElement) {
             pushTag(tag)
         }
         return shouldWriteElement
     }
 
     // For format-specific behaviour, invoked only on
-    open fun shouldWriteElement(desc: SerialDescriptor, tag: Tag, index: Int) = true
+    open fun <T: Any?> shouldWriteElement(desc: SerialDescriptor, tag: Tag, index: Int, value: T) = true
+    open fun shouldWriteNullElement(desc: SerialDescriptor, tag: Tag, index: Int) = true
```

Comment:
It's not clear that you need those (or need to change `Tagged` at all. In any case you are changing the semantics of `shouldWriteElement` which is data independent. I would say that you want the actual format to just make the appropriate decision when attempting to write the value. If the case is that a null value is written through omitting it (and this is valid for the format) that still means that it is "written".